kernel/src/log_segment.rs

```diff
@@ -422,27 +424,84 @@ impl LogSegment {
         )?))
     }
 
-    // Do a lightweight protocol+metadata log replay to find the latest Protocol and Metadata in
-    // the LogSegment
+    // Find the latest Protocol and Metadata in the LogSegment. This is done by replaying the log
+    // until the latest_crc_file/latest checkpoint is reached, at which point we can just return the
+    // P+M from the CRC directly or read from checkpoint.
+    //
+    // Note that according to the spec CRC files must contain the latest Protocol and Metadata. If
+    // we encounter a CRC file without P+M, this will return an error (and user should likely just
+    // remove such a file).
+    //
+    // Four steps:
+    // 1. If we have CRC at this version, we can just read the Protocol and Metadata from it.
+    // 2. Else, create a 'pruned' (CRC-rooted) log segment that is only the commits after the
+    //    latest CRC (if there is no CRC, we just use the whole log segment)
+    // 3. Do PM replay on the pruned log segment to find the latest Protocol/Metadata actions.
+    // 4. If no Protocol/Metadata was found, return Protocol/Metadata from CRC (if CRC doesn't
+    //    exist, we are guaranteed to get a Protocol/Metadata from checkpoint in (3)).
+    //
+    // Note that there is an optimization here: we _could_ have different types of log segments.
+    // Instead of a log segment always containing all metadata for the table, we could have a
+    // "PM-LogSegment" that is only guaranteed to contain Protocol and Metadata actions. Basically,
+    // a "PM-LogSegment" would be a log segment that is checkpoint OR CRC-rooted, whereas a full
+    // LogSegment must be checkpoint-rooted (or start from 0).
     pub(crate) fn protocol_and_metadata(
         &self,
         engine: &dyn Engine,
     ) -> DeltaResult<(Option<Metadata>, Option<Protocol>)> {
-        let actions_batches = self.replay_for_metadata(engine)?;
+        // 1. simple case if we have CRC at this version
+        if let Some(latest_crc_file) = &self.latest_crc_file {
+            if self.end_version == latest_crc_file.version {
+                let path = &latest_crc_file.location;
+                let pm = ProtocolMetadata::try_from_crc(path.clone(), engine)?;
+                return Ok((Some(pm.metadata), Some(pm.protocol)));
+            }
+        }
+
+        // 2. construct a "protocol-metadata log segment"
+        //    a) if we have a CRC file, we only need commits (CRC_version, target_version]
+        //    b) if we don't have a CRC, we proceed as normal
+        let protocol_metadata_log_segment: Cow<'_, Self> =
```

Comment:
Another option would be to add a `start_version: Option<Version>` to `replay_for_metadata`. In that case, `find_commit_cover` could quite easily just drop any versions less than `start_version`, and we wouldn't need this code really at all (and could avoid the clone)